https://www.kaggle.com/datasets/aravindpcoder/obesity-or-cvd-risk-classifyregressorcluster/ Этот набор данных включает данные для оценки уровней ожирения у лиц из Мексики, Перу и Колумбии на основе их привычек питания и физического состояния.

In [99]:
from sklearn import metrics

In [100]:
from sklearn import metrics
import pandas as pd
df = pd.read_csv("C:\\Users\\HP\\Desktop\\ObesityDataSet.csv")
df.head()

Gender   Age  Height  Weight family_history_with_overweight FAVC  FCVC  \
0  Female  21.0    1.62    64.0                            yes   no   2.0   
1  Female  21.0    1.52    56.0                            yes   no   3.0   
2    Male  23.0    1.80    77.0                            yes   no   2.0   
3    Male  27.0    1.80    87.0                             no   no   3.0   
4    Male  22.0    1.78    89.8                             no   no   2.0   

   NCP       CAEC SMOKE  CH2O  SCC  FAF  TUE        CALC  \
0  3.0  Sometimes    no   2.0   no  0.0  1.0          no   
1  3.0  Sometimes   yes   3.0  yes  3.0  0.0   Sometimes   
2  3.0  Sometimes    no   2.0   no  2.0  1.0  Frequently   
3  3.0  Sometimes    no   2.0   no  2.0  0.0  Frequently   
4  1.0  Sometimes    no   2.0   no  0.0  0.0   Sometimes   

                  MTRANS           NObeyesdad  
0  Public_Transportation        Normal_Weight  
1  Public_Transportation        Normal_Weight  
2  Public_Transportation        Normal_Weight  
3                Walking   Overweight_Level_I  
4  Public_Transportation  Overweight_Level_II

Для дальнейшей работы выберем некоторые атрибуты

In [101]:
df = df[['Weight', 'CH2O', 'FAF', 'TUE']]
df.head()

Weight  CH2O  FAF  TUE
0    64.0   2.0  0.0  1.0
1    56.0   3.0  3.0  0.0
2    77.0   2.0  2.0  1.0
3    87.0   2.0  2.0  0.0
4    89.8   2.0  0.0  0.0

Для выполнения данного задания вам необходимо сначала оценить, насколько набор
данных, выбранный вами ранее, подходит для решения задачи. Для этого следует:

1) Оценить какой атрибут выступит в роли целевого класса (метки)
2) Оценить сбалансированность классов

In [102]:
import pandas as pd
import numpy as np

def Obesity_values(row):
    first = row["Weight"]

    if (first < 18.5):
        return 'Normal'
    elif (18.5 < first < 24.9):
        return 'Normal'
    elif (25.0 < first < 29.9):
        return 'Overweight'
    elif (30.0 < first < 34.9):
        return 'Obesity I '
    elif (35.0< first < 39.9):
        return 'Obesity II'
    else:
        return 'Obesity III'
        

df["Obesity_values"] = df.apply(lambda row: Obesity_values(row), axis=1)
df.head()

Weight  CH2O  FAF  TUE Obesity_values
0    64.0   2.0  0.0  1.0    Obesity III
1    56.0   3.0  3.0  0.0    Obesity III
2    77.0   2.0  2.0  1.0    Obesity III
3    87.0   2.0  2.0  0.0    Obesity III
4    89.8   2.0  0.0  0.0    Obesity III

В роли целевого атрибута выступит "Obesity_values"

In [103]:
df.shape

(2111, 5)

In [104]:
df['Obesity_values'].value_counts()

Obesity_values
Obesity III    2106
Obesity II        5
Name: count, dtype: int64

• Недостаточный вес Менее 18,5 
• Нормальный от 18,5 до 24,9 
• Избыточный вес от 25,0 до 29,9 
• Ожирение I от 30,0 до 34,9 
• Ожирение II от 35,0 до 39,9 
• Ожирение III Выше 40.

можно увидеть что в основном это люди с ожирением 3 степени, класс несбалансированный

1. Необходимо оценить и сравнить результаты классификации, используя следующие
алгоритмы классификации:
* kNN
* дерево решений

kNN

In [105]:
import numpy as np 
import matplotlib.pyplot as plt 
import pandas as pd


Переменная x содержит первые 6 столбца набора данных (т.е. атрибуты), а y содержит метки.

In [106]:
X = df.iloc[:, :-1].values 
y = df.iloc[:, 4].values


скрипт разбивает набор данных на 80% данных поезда и 20% данных теста


In [107]:
from sklearn.model_selection import train_test_split 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)


масштабирование функций


In [108]:
from sklearn.preprocessing import StandardScaler 
scaler = StandardScaler() 
scaler.fit(X_train) 
X_train = scaler.transform(X_train) 
X_test = scaler.transform(X_test)


импортируем класс KNeighborsClassifier из библиотеки sklearn.neighbors. Во второй строке этот класс инициализируется одним параметром, то есть n_neigbours. По сути, это значение для K. Идеального значения для K не существует, и оно выбирается после тестирования и оценки, однако для начала наиболее часто используемым значением для алгоритма KNN является 5


In [109]:
from sklearn.neighbors import KNeighborsClassifier 
classifier = KNeighborsClassifier(n_neighbors=5) 
classifier.fit(X_train, y_train)



KNeighborsClassifier()

In [110]:
y_pred = classifier.predict(X_test)


Для задач классификации часто используются такие показатели, как матрица неточностей, точность, отзыв и оценка F1.Библиотека метрик Scikit = -Learn содержит методы классификации_report и confusion_matrix, которые можно использовать для расчета этих показателей для нас:


Добавила zero_division=0, тк датасет содержит классы, которые не были предсказаны моделью

In [124]:
from sklearn.metrics import classification_report, confusion_matrix 
print(confusion_matrix(y_test, y_pred)) 
print(classification_report(y_test, y_pred, zero_division=0))


[[  0   2]
 [  0 421]]
              precision    recall  f1-score   support

  Obesity II       0.00      0.00      0.00         2
 Obesity III       1.00      1.00      1.00       421

    accuracy                           1.00       423
   macro avg       0.50      0.50      0.50       423
weighted avg       0.99      1.00      0.99       423




Obesity II:
Если F1-score (также известный как F-мера) равен нулю, это означает, что модель не смогла правильно предсказать ни одного положительного экземпляра из класса. F1-мера является мерой качества классификации, которая учитывает как точность, так и полноту модели. Если модель не может правильно предсказать ни один положительный экземпляр, то ее точность и полнота будут равны нулю, что приведет к F1-мере равной нулю.

1) вывод по weighted avg: модель правильно классифицировала все объекты в обучающей выборке с точностью 0.99. Однако, поскольку это только обучающая выборка, реальная точность модели на всех данных может быть ниже.
2) вывод по macro avg: точность классификации для каждого класса составляет 50%. Это может указывать на то, что модель работает плохо и нуждается в доработке.
3) Если accuracy равен 1.0, то это означает, что модель точно предсказала все метки классов в обучающей выборке.

Дерево решений

In [120]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
%matplotlib inline


в нашем наборе данных 2111 записи и 5 атрибутов


In [121]:
df.shape

(2111, 5)

In [123]:
X = df.drop('Obesity_values', axis=1) 
y = df['Obesity_values']


Здесь переменная X содержит все столбцы из набора данных, кроме столбца «Obesity_values», который является меткой. Переменная y содержит значения из столбца «Obesity_values». Переменная X – это наш набор атрибутов, а переменная y содержит соответствующие метки.


In [ ]:
from sklearn.model_selection import train_test_split 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)


разделяем наши данные на обучающие и тестовые наборы
В приведенном выше коде параметр test_size указывает соотношение набора тестов, которое мы используем для разделения 20% данных на набор тестов и 80% для обучения.


обучаем алгоритм дерева решений на этих данных и сделаем прогнозы


мы будем использовать класс DecisionTreeClassifier. Метод подгонки этого класса вызывается для обучения алгоритма на обучающих данных, которые передаются в качестве параметра методу подгонки.

In [ ]:
from sklearn.tree import DecisionTreeClassifier 
classifier = DecisionTreeClassifier() 
classifier.fit(X_train, y_train)


In [126]:
y_pred = classifier.predict(X_test)


Добавила zero_division=0, тк датасет содержит классы, которые не были предсказаны моделью

In [127]:
from sklearn.metrics import classification_report, confusion_matrix 
print(confusion_matrix(y_test, y_pred)) 
print(classification_report(y_test, y_pred, zero_division=0))


[[  0   2]
 [  0 421]]
              precision    recall  f1-score   support

  Obesity II       0.00      0.00      0.00         2
 Obesity III       1.00      1.00      1.00       421

    accuracy                           1.00       423
   macro avg       0.50      0.50      0.50       423
weighted avg       0.99      1.00      0.99       423



Результаты совпали с алгоритмом KNN

Общий вывод: 
1) Хочется заметить в данном датасете имеется метка, в нашем случае это атрибут "Obesity_values", но при анализе получили такой результат Obesity III = 2106, Obesity II = 5, что сразу указывает на несбалансированность классов, это можно было бы решить с помощью модуля NEARMISS
2) Опираясь на п.1 класс Obesity II в итоге не был предсказан моделью, из за чего в итоговом результате precision    recall  f1-score равны нулю
3) в 1 и 2 алгоритме класс Obesity III получил в результате f1-score означает, что модель идеально классифицировала все данные